File: backend-python/app/models/payment.py
"""Production payment, wallet, settlement and admin models — Phase 5 · Sprint 5.6.

These mirror `shared/src/types/payment.ts` field-for-field (camelCase JSON) so
the frontend needs zero changes when it is pointed at this FastAPI service
instead of the mock router (`backend/src/mock/payment-routes.ts`).
"""

from __future__ import annotations

from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

PaymentGateway = Literal["razorpay", "wallet", "cod", "mixed"]
GatewayPaymentStatus = Literal[
    "created",
    "authorized",
    "captured",
    "paid",
    "failed",
    "refunded",
    "partially_refunded",
    "cancelled",
]
LedgerDirection = Literal["credit", "debit"]
LedgerReason = Literal[
    "order-payment",
    "wallet-topup",
    "refund",
    "cashback",
    "settlement",
    "withdrawal",
    "incentive",
    "adjustment",
    "penalty",
]
LedgerStatus = Literal["success", "pending", "failed"]
GatewayRefundStatus = Literal["requested", "approved", "processing", "processed", "rejected", "failed"]
SettlementStatus = Literal["pending", "approved", "processing", "settled", "rejected"]
WithdrawalStatus = Literal["requested", "approved", "processing", "paid", "rejected"]
AccountRole = Literal["customer", "partner", "rider", "admin"]


class RazorpayConfig(BaseModel):
    keyId: str
    enabled: bool
    currency: str = "INR"
    mode: Literal["test", "live", "disabled"] = "test"


class RazorpayOrderResult(BaseModel):
    ok: bool = True
    paymentId: str
    gatewayOrderId: str
    keyId: str
    currency: str = "INR"
    amount: float
    walletApplied: float
    payableAmount: float
    amountInPaise: int
    fullyPaidByWallet: bool
    receipt: str
    notes: Dict[str, str] = Field(default_factory=dict)


class GatewayPayment(BaseModel):
    id: str
    orderId: Optional[str] = None
    accountId: str
    gateway: PaymentGateway
    status: GatewayPaymentStatus
    amount: float
    walletAmount: float = 0
    gatewayAmount: float = 0
    currency: str = "INR"
    gatewayOrderId: Optional[str] = None
    gatewayPaymentId: Optional[str] = None
    signatureVerified: bool = False
    purpose: str = "Order payment"
    failureReason: Optional[str] = None
    refundedAmount: float = 0
    createdAt: str
    updatedAt: str


class RazorpaySuccessPayload(BaseModel):
    paymentId: str
    razorpay_order_id: Optional[str] = None
    razorpay_payment_id: Optional[str] = None
    razorpay_signature: Optional[str] = None


class SimulateCheckoutPayload(BaseModel):
    gatewayOrderId: str


class PaymentVerificationResult(BaseModel):
    ok: bool
    verified: bool
    message: str
    payment: GatewayPayment


class CreateOrderPayload(BaseModel):
    amount: float = Field(gt=0, le=1_000_000)
    orderId: Optional[str] = None
    purpose: str = "Order payment"
    walletAmount: float = 0
    savedMethodId: Optional[str] = None


class PaymentFailurePayload(BaseModel):
    paymentId: str
    reason: Optional[str] = None
    code: Optional[str] = None


class RefundRequestPayload(BaseModel):
    amount: Optional[float] = None
    reason: str = "Customer requested a refund"
    destination: Literal["source", "wallet"] = "source"


class WalletLedgerEntry(BaseModel):
    id: str
    accountId: str
    role: Literal["customer", "partner", "rider"]
    direction: LedgerDirection
    reason: LedgerReason
    amount: float
    balanceAfter: float
    currency: str = "INR"
    reference: Optional[str] = None
    orderId: Optional[str] = None
    paymentId: Optional[str] = None
    note: str = ""
    status: LedgerStatus = "success"
    createdAt: str
    dateLabel: str


class WalletLedgerResult(BaseModel):
    entries: List[WalletLedgerEntry] = Field(default_factory=list)
    balance: float = 0
    pending: float = 0
    lifetimeCredit: float = 0
    lifetimeDebit: float = 0
    currency: str = "INR"


class WalletCreditPayload(BaseModel):
    reason: LedgerReason = "wallet-topup"
    amount: float = Field(gt=0)
    note: str = "Wallet credited"
    reference: Optional[str] = None


class WalletTopupPayload(BaseModel):
    amount: float = Field(gt=0)


class GatewayRefund(BaseModel):
    id: str
    paymentId: str
    orderId: Optional[str] = None
    accountId: str
    amount: float
    reason: str
    status: GatewayRefundStatus
    gatewayRefundId: Optional[str] = None
    destination: Literal["source", "wallet"] = "source"
    createdAt: str
    updatedAt: str
    dateLabel: str
    timeline: List[Dict[str, str]] = Field(default_factory=list)


class Settlement(BaseModel):
    id: str
    accountId: str
    role: Literal["partner", "rider"]
    periodStart: str
    periodEnd: str
    periodLabel: str
    orders: int
    grossAmount: float
    commission: float
    taxDeducted: float
    incentives: float
    netAmount: float
    status: SettlementStatus
    utr: Optional[str] = None
    settledAt: Optional[str] = None
    createdAt: str


class SettlementResult(BaseModel):
    items: List[Settlement] = Field(default_factory=list)
    totalSettled: float = 0
    totalPending: float = 0
    currency: str = "INR"


class WithdrawalRequest(BaseModel):
    id: str
    accountId: str
    role: Literal["partner", "rider"]
    amount: float
    method: Literal["bank", "upi"]
    destination: str
    status: WithdrawalStatus
    requestedAt: str
    processedAt: Optional[str] = None
    rejectionReason: Optional[str] = None
    reference: Optional[str] = None
    dateLabel: str


class WithdrawalResult(BaseModel):
    items: List[WithdrawalRequest] = Field(default_factory=list)
    available: float = 0
    pendingAmount: float = 0
    minimumAmount: float = 100
    currency: str = "INR"


class CreateWithdrawalPayload(BaseModel):
    amount: float = Field(gt=0)
    method: Literal["bank", "upi"] = "bank"
    destination: str = "Primary bank account"


class EarningsBreakdown(BaseModel):
    today: float = 0
    week: float = 0
    month: float = 0
    lifetime: float = 0
    orders: int = 0
    averagePerOrder: float = 0
    commissionRate: float = 0.18
    pendingSettlement: float = 0
    currency: str = "INR"
    series: List[Dict[str, float]] = Field(default_factory=list)


class RiderIncentive(BaseModel):
    id: str
    title: str
    description: str
    target: int
    progress: int
    reward: float
    status: Literal["active", "completed", "expired"]
    expiresAt: str


class AdminPaymentKpis(BaseModel):
    grossVolume: float = 0
    successfulPayments: int = 0
    failedPayments: int = 0
    successRate: float = 100
    refundedAmount: float = 0
    refundCount: int = 0
    walletFloat: float = 0
    pendingSettlements: int = 0
    pendingWithdrawals: int = 0
    currency: str = "INR"


class AdminPaymentDashboard(BaseModel):
    kpis: AdminPaymentKpis
    recentPayments: List[GatewayPayment] = Field(default_factory=list)
    gatewayMode: Literal["test", "live", "disabled"] = "test"
    series: List[Dict[str, float]] = Field(default_factory=list)


class AdminWalletMonitorRow(BaseModel):
    accountId: str
    name: str
    role: Literal["customer", "partner", "rider"]
    balance: float
    pending: float
    lifetimeCredit: float
    lifetimeDebit: float
    lastActivityAt: Optional[str] = None
    flagged: bool = False


class ApproveSettlementPayload(BaseModel):
    utr: Optional[str] = None


class RejectReasonPayload(BaseModel):
    reason: str = ""


class ApproveWithdrawalPayload(BaseModel):
    reference: Optional[str] = None

File: customer-frontend/src/shared/types/payment.ts
/**
 * Phase 5 · Sprint 5.6 — Production payment, wallet, settlement contracts.
 *
 * These types are the single source of truth shared by:
 *   • the frontend data layer (`backend/src/payments/*`)
 *   • the mock backend (`backend/src/mock/payments-core.ts`)
 *   • the FastAPI service (`backend-python/app/models/payment.py`)
 *
 * Additive only — no existing type was changed.
 */

export type PaymentGateway = "razorpay" | "wallet" | "cod" | "mixed";

export type GatewayPaymentStatus =
  | "created"
  | "authorized"
  | "captured"
  | "paid"
  | "failed"
  | "refunded"
  | "partially_refunded"
  | "cancelled";

export type RazorpayConfig = {
  /** Publishable key id (rzp_test_… / rzp_live_…). Never the secret. */
  keyId: string;
  enabled: boolean;
  currency: string;
  /** "test" | "live" — derived from the key prefix on the server. */
  mode: "test" | "live" | "disabled";
};

/** Response of POST /api/payments/razorpay/order */
export type RazorpayOrderResult = {
  ok: boolean;
  /** Internal payment document id. */
  paymentId: string;
  /** Razorpay order id (order_…) — empty when the whole amount came from wallet. */
  gatewayOrderId: string;
  keyId: string;
  currency: string;
  /** Total order amount in rupees. */
  amount: number;
  /** Portion settled from the wallet (mixed payment). */
  walletApplied: number;
  /** Portion Razorpay must collect, in rupees. */
  payableAmount: number;
  /** Razorpay expects paise. */
  amountInPaise: number;
  /** True when wallet covered everything and no checkout is required. */
  fullyPaidByWallet: boolean;
  receipt: string;
  notes: Record<string, string>;
};

/** Checkout handler payload handed back by Razorpay. */
export type RazorpaySuccessPayload = {
  razorpay_payment_id: string;
  razorpay_order_id: string;
  razorpay_signature: string;
};

export type PaymentVerificationResult = {
  ok: boolean;
  verified: boolean;
  message: string;
  payment: GatewayPayment;
};

export type GatewayPayment = {
  id: string;
  orderId: string | null;
  accountId: string;
  gateway: PaymentGateway;
  status: GatewayPaymentStatus;
  amount: number;
  walletAmount: number;
  gatewayAmount: number;
  currency: string;
  gatewayOrderId: string | null;
  gatewayPaymentId: string | null;
  signatureVerified: boolean;
  purpose: string;
  failureReason: string | null;
  refundedAmount: number;
  createdAt: string;
  updatedAt: string;
};

/* ------------------------------ wallet ledger ----------------------------- */

export type LedgerDirection = "credit" | "debit";

export type LedgerReason =
  | "order-payment"
  | "wallet-topup"
  | "refund"
  | "cashback"
  | "settlement"
  | "withdrawal"
  | "incentive"
  | "adjustment"
  | "penalty";

export type WalletLedgerEntry = {
  id: string;
  accountId: string;
  role: "customer" | "partner" | "rider";
  direction: LedgerDirection;
  reason: LedgerReason;
  amount: number;
  balanceAfter: number;
  currency: string;
  reference: string | null;
  orderId: string | null;
  paymentId: string | null;
  note: string;
  status: "success" | "pending" | "failed";
  createdAt: string;
  dateLabel: string;
};

export type WalletLedgerResult = {
  entries: WalletLedgerEntry[];
  balance: number;
  pending: number;
  lifetimeCredit: number;
  lifetimeDebit: number;
  currency: string;
};

/* -------------------------------- refunds -------------------------------- */

export type GatewayRefundStatus = "requested" | "approved" | "processing" | "processed" | "rejected" | "failed";

export type GatewayRefund = {
  id: string;
  paymentId: string;
  orderId: string | null;
  accountId: string;
  amount: number;
  reason: string;
  status: GatewayRefundStatus;
  /** rfnd_… once Razorpay accepted it. */
  gatewayRefundId: string | null;
  /** "wallet" refunds land instantly; "source" goes back to the card/UPI. */
  destination: "source" | "wallet";
  createdAt: string;
  updatedAt: string;
  dateLabel: string;
  timeline: { label: string; at: string }[];
};

/* ------------------------------ settlements ------------------------------- */

export type SettlementStatus = "pending" | "approved" | "processing" | "settled" | "rejected";

export type Settlement = {
  id: string;
  accountId: string;
  role: "partner" | "rider";
  periodStart: string;
  periodEnd: string;
  periodLabel: string;
  orders: number;
  grossAmount: number;
  commission: number;
  taxDeducted: number;
  incentives: number;
  netAmount: number;
  status: SettlementStatus;
  utr: string | null;
  settledAt: string | null;
  createdAt: string;
};

export type SettlementResult = {
  items: Settlement[];
  totalSettled: number;
  totalPending: number;
  currency: string;
};

/* ------------------------------ withdrawals ------------------------------- */

export type WithdrawalStatus = "requested" | "approved" | "processing" | "paid" | "rejected";

export type WithdrawalRequest = {
  id: string;
  accountId: string;
  role: "partner" | "rider";
  amount: number;
  method: "bank" | "upi";
  destination: string;
  status: WithdrawalStatus;
  requestedAt: string;
  processedAt: string | null;
  rejectionReason: string | null;
  reference: string | null;
  dateLabel: string;
};

export type WithdrawalResult = {
  items: WithdrawalRequest[];
  available: number;
  pendingAmount: number;
  minimumAmount: number;
  currency: string;
};

/* -------------------------------- earnings -------------------------------- */

export type EarningsBreakdown = {
  today: number;
  week: number;
  month: number;
  lifetime: number;
  orders: number;
  averagePerOrder: number;
  commissionRate: number;
  pendingSettlement: number;
  currency: string;
  series: { label: string; amount: number }[];
};

export type RiderIncentive = {
  id: string;
  title: string;
  description: string;
  target: number;
  progress: number;
  reward: number;
  status: "active" | "completed" | "expired";
  expiresAt: string;
};

/* -------------------------------- admin ----------------------------------- */

export type AdminPaymentKpis = {
  grossVolume: number;
  successfulPayments: number;
  failedPayments: number;
  successRate: number;
  refundedAmount: number;
  refundCount: number;
  walletFloat: number;
  pendingSettlements: number;
  pendingWithdrawals: number;
  currency: string;
};

export type AdminPaymentDashboard = {
  kpis: AdminPaymentKpis;
  recentPayments: GatewayPayment[];
  gatewayMode: "test" | "live" | "disabled";
  series: { label: string; captured: number; refunded: number }[];
};

export type AdminWalletMonitorRow = {
  accountId: string;
  name: string;
  role: "customer" | "partner" | "rider";
  balance: number;
  pending: number;
  lifetimeCredit: number;
  lifetimeDebit: number;
  lastActivityAt: string | null;
  flagged: boolean;
};
